infrastructure/src/main/java/org/corfudb/infrastructure/log/FileChannelPerf.java

```diff
@@ -0,0 +1,183 @@
+        detectorEnabled = report;
+        spikeLatency = maxLatencyVal;
+        readMetrics = ServerContext.getMetrics ().timer (READ_METRICS);
+        writeMetrics = ServerContext.getMetrics ().timer (WRITE_METRICS);
+        syncMetrics = ServerContext.getMetrics ().timer (SYNC_METRICS);
+        slidingWindow = new SlidingWindow(cntSpike, cntSpike, maxLatencyVal);
+    }
+
+    static private long getUnit(long size) {
+        return (size + MB_DATA - 1) / MB_DATA;
+    }
+
+    public FileChannelPerf(FileChannel fc) {
+        fileChannel = fc;
+    }
+
+    public static FileChannelPerf open(Path path, Set<? extends OpenOption> options,
+             FileAttribute<?>... attrs) throws IOException {
+        FileChannel fileChannel = FileChannel.open(path, options, attrs);
+        return new FileChannelPerf(fileChannel);
+    }
+
+    public static FileChannelPerf open(Path path, OpenOption... options) throws IOException {
+        FileChannel fileChannel = FileChannel.open(path, options);
+        return new FileChannelPerf(fileChannel);
+    }
+
+    public long size() throws IOException {
+        return fileChannel.size ();
+    }
+
+    public long position() throws IOException {
+        return fileChannel.position();
+    }
+
+    public FileChannelPerf position(long address) throws IOException {
+        fileChannel.position(address);
+        return this;
+    }
+
+    public int read(ByteBuffer buf, boolean use_position, long position) throws IOException {
+        long start = System.nanoTime ();
+        int size = buf.remaining ();
+        int res = 0;
+
+        if (!use_position) {
+            res = fileChannel.read (buf);
+        } else {
+            res = fileChannel.read (buf, position);
+        }
+        update(readMetrics, READ_METRICS, start, size);
+        return res;
+    }
+
+    public int read(ByteBuffer dst) throws IOException {
+        return read (dst, false, 0);
+    }
+
+    public int read(ByteBuffer dst, long position) throws IOException {
+        return read (dst, true, position);
+    }
+
+    public int write(ByteBuffer buf, boolean use_position, long position) throws IOException {
+        long start = System.nanoTime ();
+        int size = buf.remaining ();
+        int res = 0;
+        if (!use_position)
+            res = fileChannel.write (buf);
+        else {
+            res = fileChannel.write (buf, position);
+        }
+        update(writeMetrics, WRITE_METRICS, start, size);
+        return res;
+    }
+
+    public int write(ByteBuffer dst) throws IOException {
+        return write (dst, false, 0);
+    }
+
+
+    public int write(ByteBuffer dst, long position) throws IOException {
+        return write (dst, true, position);
+    }
+
+    public void force(boolean force) throws IOException {
+        long start = System.nanoTime ();
+        fileChannel.force (force);
+        update(syncMetrics, SYNC_METRICS, start, MB_DATA);
```

Comment:
You probably want to measure non force and force separately right otherwise the histogram would be incorrect as it is measuring both metadata fsync and non-metadata fsync (fdatasync) right?

Reply:
@hisundar there is no histogram. I think the intention here is to detect any read/write spike within a time window. We need something for IO stats though, but this way (at this layer) of detection isn't the best way to do that.

Reply:
Oh, I see. You are referring to the timer.  I think there should be histograms on the handler, but not at this level. For low level IO, we can just collect disk stats directly.